<a href="https://colab.research.google.com/github/KravtsovaAlyona/data_science_platform/blob/main/%D0%9A%D1%80%D0%B0%D0%B2%D1%86%D0%BE%D0%B2%D0%B0_6_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science – Упражнения для обработки текста


ФИО: Кравцова А.Е

Почта: KravtsovaAE@mgpu.ru

## NLP

### Упражнение 1: Часто встречающиеся слова
Найдите в «Моби Дике» наиболее часто используемые слова, которые не являются стоп-словами и знаками препинания. Подсказка: здесь может пригодиться [`str.isalpha()`](https://docs.python.org/3/library/stdtypes.html#str.isalpha).

In [1]:
#импорт библиотеки, текстов и стоп-слов
import nltk
nltk.download('all')
from nltk.corpus import stopwords
stopwords = nltk.corpus.stopwords.words('english')
from nltk.book import *

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_ru.zip.
[nltk_data]    | Downloading package basque_grammars to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping grammars/basque_grammars.zip.
[nltk_data]    | Downloading package bcp47 to /root/nltk_data...
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   U

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


In [22]:
frequency = FreqDist(text1)
print(frequency)
print('\n',frequency.most_common(50))

<FreqDist with 19317 samples and 260819 outcomes>

 [(',', 18713), ('the', 13721), ('.', 6862), ('of', 6536), ('and', 6024), ('a', 4569), ('to', 4542), (';', 4072), ('in', 3916), ('that', 2982), ("'", 2684), ('-', 2552), ('his', 2459), ('it', 2209), ('I', 2124), ('s', 1739), ('is', 1695), ('he', 1661), ('with', 1659), ('was', 1632), ('as', 1620), ('"', 1478), ('all', 1462), ('for', 1414), ('this', 1280), ('!', 1269), ('at', 1231), ('by', 1137), ('but', 1113), ('not', 1103), ('--', 1070), ('him', 1058), ('from', 1052), ('be', 1030), ('on', 1005), ('so', 918), ('whale', 906), ('one', 889), ('you', 841), ('had', 767), ('have', 760), ('there', 715), ('But', 705), ('or', 697), ('were', 680), ('now', 646), ('which', 640), ('?', 637), ('me', 627), ('like', 624)]


Самые частые элементы текста - знаки припенания и мешающие, бесполезные слова - местоимения, предлоги, служебные глаголы и тд. Данные слова портят анализ текста, поэтому отфильтрую их

In [4]:
frequency_list = FreqDist(word.lower() for word in text1 if word.isalpha()
  and word.lower() not in stopwords)
frequency_list.most_common(50)

[('whale', 1226),
 ('one', 921),
 ('like', 647),
 ('upon', 566),
 ('man', 527),
 ('ship', 518),
 ('ahab', 511),
 ('ye', 472),
 ('sea', 455),
 ('old', 450),
 ('would', 432),
 ('though', 384),
 ('head', 345),
 ('yet', 345),
 ('boat', 336),
 ('time', 334),
 ('long', 333),
 ('captain', 329),
 ('still', 312),
 ('great', 306),
 ('said', 304),
 ('two', 298),
 ('must', 283),
 ('seemed', 283),
 ('white', 281),
 ('last', 277),
 ('see', 272),
 ('thou', 271),
 ('way', 271),
 ('whales', 268),
 ('stubb', 257),
 ('queequeg', 252),
 ('little', 249),
 ('round', 247),
 ('three', 245),
 ('say', 244),
 ('men', 244),
 ('sperm', 244),
 ('may', 240),
 ('first', 235),
 ('every', 232),
 ('well', 230),
 ('us', 228),
 ('much', 223),
 ('could', 216),
 ('good', 216),
 ('hand', 214),
 ('side', 208),
 ('ever', 206),
 ('never', 206)]

 Итак, самым популярным словом является слово "whale" - кит (1226 раз), далее "one", "like", "upon" - эти слова также не несут значимую смысловую нагрузку, поэтому можно самостоятельно добавить их в список стоп-слов

In [5]:
stopwords.extend(['one','like','upon'])

In [6]:
frequency_list_final = FreqDist(word.lower() for word in text1 if word.isalpha()
  and word.lower() not in stopwords)
frequency_list_final.most_common(50)

[('whale', 1226),
 ('man', 527),
 ('ship', 518),
 ('ahab', 511),
 ('ye', 472),
 ('sea', 455),
 ('old', 450),
 ('would', 432),
 ('though', 384),
 ('head', 345),
 ('yet', 345),
 ('boat', 336),
 ('time', 334),
 ('long', 333),
 ('captain', 329),
 ('still', 312),
 ('great', 306),
 ('said', 304),
 ('two', 298),
 ('must', 283),
 ('seemed', 283),
 ('white', 281),
 ('last', 277),
 ('see', 272),
 ('thou', 271),
 ('way', 271),
 ('whales', 268),
 ('stubb', 257),
 ('queequeg', 252),
 ('little', 249),
 ('round', 247),
 ('three', 245),
 ('say', 244),
 ('men', 244),
 ('sperm', 244),
 ('may', 240),
 ('first', 235),
 ('every', 232),
 ('well', 230),
 ('us', 228),
 ('much', 223),
 ('could', 216),
 ('good', 216),
 ('hand', 214),
 ('side', 208),
 ('ever', 206),
 ('never', 206),
 ('look', 205),
 ('starbuck', 198),
 ('deck', 196)]

Теперь второе по частоте встречаемости слово - человек (527 раз), третье - корабль (518 раз)

#Упражнение 2
Вы злой спамер, который заметил, что многие люди пытаются запутать свою электронную почту, используя пометку: «alex at utah dot edu». Ниже приведены три примера текста таких писем. Попробуйте извлечь «alex at utah dot edu» и т. д. Начните с первой строки. Затем расширьте свое регулярное выражение, чтобы оно работало со всеми ними одновременно. Обратите внимание, что второе и третье сделать намного сложнее!

You're an evil Spammer who's observed that many people try to obfuscate their e-mail using this notation: "alex at utah dot edu". Below are three examples of such e-mails text. Try to extract "alex at utah dot edu", etc. Start with the first string. Then extend your regular expression to work on all of them at the same time. Note that the second and third are slightly harder to do!

In [7]:
import re
html_smart = "You can reach me: alex at utah dot edu"
html_smart2 = "You can reach me: alex dot lex at utah dot edu"
html_smart3 = "You can reach me: alex dot lex at sci dot utah dot edu"

In [8]:
def testRegex(regex):
    for html in (html_smart, html_smart2, html_smart3):
        print(re.search(regex, html).group())

In [38]:
mail_regex=r"alex(?: [\w\s\.]+)? at (?:[\w\s\.]+ dot )?utah dot edu"

In [18]:
testRegex(mail_regex)

alex at utah dot edu
alex dot lex at utah dot edu
alex dot lex at sci dot utah dot edu


## Упражнение 2.2: Найдите наречия

Напишите регулярное выражение, которое находит все наречия в предложении. Наречия характеризуются окончанием на «ly».

In [19]:
text2_2 = "He was carefully disguised but captured quickly by police."

In [20]:
adverb_re = r'\w+ly'
re.findall(adverb_re, text2_2)

['carefully', 'quickly']

### Упражнение 2.3: Телефонные номера

Извлеките из текста номера телефонов, которые следуют шаблону (xxx) xxx-xxxx:






In [12]:
phone_numbers = "(857) 131-2235, (801) 134-2215, but this one (12) 13044441 shouldnt match. Also, this is common in twelve (12) countries and one (1) state"

In [37]:
phone_re = r'\(\d{3}\)\s\d{3}-\d{4}'
re.findall(phone_re, phone_numbers)

['(857) 131-2235', '(801) 134-2215']

### Упражнение 2.4. HTML-контент

Извлеките содержимое между тегами `<b>` и `<i>`, но не другими тегами:

In [14]:
html_tags = "This is <b>important</b> and <u>very</u><i>timely</i>"

In [40]:
tags_re = re.findall(r'<b>(.*?)</b>.*?<i>(.*?)</i>', html_tags)
print(tags_re)

[('important', 'timely')]
